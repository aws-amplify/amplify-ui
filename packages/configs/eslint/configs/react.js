module.exports = {
  plugins: ['react', 'react-hooks'],
  extends: [
    'plugin:react/recommended',
    'plugin:react-hooks/recommended',

    // extend internal config last to ensure prettier is final loaded extension
    '@aws-amplify/amplify-ui',
  ],
  parserOptions: { ecmaFeatures: { jsx: true } },
  settings: { react: { version: 'detect' } },
  rules: {
    // react rules either not in recommended rule set or overridden
    'react/destructuring-assignment': ['error', 'always'],
    'react/jsx-boolean-value': 'error',
    'react/jsx-no-constructed-context-values': 'error',
    'react/jsx-no-useless-fragment': ['error', { allowExpressions: true }],
    'react/jsx-props-no-spreading': 'off',
    'react/jsx-wrap-multilines': ['error', { declaration: 'ignore' }],
    'react/no-array-index-key': 'off',
    'react/no-danger': 'error',
    'react/no-unused-prop-types': 'error',
    'react/prop-types': 'off',
    'react/static-property-placement': ['error', 'static public field'],

    // react hook rules
    'react-hooks/rules-of-hooks': 'error',
    'react-hooks/exhaustive-deps': 'error',
  },
};
