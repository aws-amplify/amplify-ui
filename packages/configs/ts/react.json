{
  "extends": "./base.json",
  "compilerOptions": {
    "jsx": "react"
  }
}
