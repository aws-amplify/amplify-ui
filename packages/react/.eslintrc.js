// TODO remove these once the full set of rules can be turned on for the repo
// extensions that should be ran against the entire repo
const sharedExtensions = [
  'plugin:react-hooks/recommended',

  // always extend last to override previous extensions
  'prettier',
];

// rules that should be ran against the entire repo
const sharedRules = {
  // react hook rules
  'react-hooks/rules-of-hooks': 'error',
  'react-hooks/exhaustive-deps': 'error',
};

module.exports = {
  env: { node: true },
  root: true,
  ignorePatterns: [
    // top level directories
    '__tests__',
    'dist',
    'node_modules',
    'scripts',

    // config files
    '.eslintrc.js',
    'jest.config.js',
    'jest.setup.ts',
    'rollup.config.js',
    'tsup.config.ts',

    // TODO remove once the full set of rules can be turned on for the repo
    'src/react-shim.js',
  ],
  extends: sharedExtensions,
  plugins: ['react-hooks'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaFeatures: { jsx: true },
    ecmaVersion: 12,
    project: ['./tsconfig.json'],
    tsconfigRootDir: __dirname,
    sourceType: 'module',
  },
  overrides: [
    {
      files: ['**/*.spec.*', '**/*.test.*'],
      parser: '@typescript-eslint/parser',
      parserOptions: {
        project: ['./tsconfig.dev.json'],
      },
      plugins: ['@typescript-eslint', 'jest'],
      rules: {
        // Turn the original rule off for test files, turn on the jest rule
        '@typescript-eslint/unbound-method': 'off',
        'jest/unbound-method': 'error',
      },
    },
    // TODO remove once the full set of rules can be turned on for the repo
    // To keep the rules-of-hooks on for the entire ui-react package while
    // we add the larger set of rules incrementally, use an overrides object
    // to incrementally apply the new rules
    {
      files: [
        // TODO Uncomment the below lines on an individual basis as the updates to pass linting are completed.
        // If needed, these can be broken down further, e.g. 'src/primitives/Alert/**/*'
        'src/components/**/*',
        'src/helpers/**/*',
        'src/hooks/**/*',
        'src/primitives/+(shared|utils)/**/*',
        'src/studio',
        // 'src/primitives/**/*',
      ],
      extends: [
        'eslint:recommended',
        'plugin:jest/recommended',
        'plugin:react/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',

        // extend last to override previous extensions
        ...sharedExtensions,
      ],
      plugins: ['@typescript-eslint', 'jest', 'react', 'react-hooks'],
      settings: {
        react: {
          version: 'detect',
        },
      },
      rules: {
        ...sharedRules,

        // typescript eslint rules either not in recommended rule set or overridden
        '@typescript-eslint/explicit-module-boundary-types': 2,
        '@typescript-eslint/member-ordering': 'error',
        '@typescript-eslint/no-extra-semi': 'error',
        '@typescript-eslint/no-floating-promises': ['off'],
        '@typescript-eslint/no-unused-expressions': [
          'error',
          { allowTernary: true },
        ],
        '@typescript-eslint/no-use-before-define': ['error'],
        '@typescript-eslint/no-unused-vars': [
          'error',
          { argsIgnorePattern: '_', varsIgnorePattern: '_' },
        ],
        '@typescript-eslint/prefer-nullish-coalescing': 'error',
        '@typescript-eslint/restrict-template-expressions': ['off'],
        '@typescript-eslint/unbound-method': 'error',

        // eslint rules either not in recommended rule set or overridden
        'comma-dangle': ['error', 'only-multiline'],
        'function-paren-newline': 'off',
        'generator-star-spacing': 'off',
        'global-require': 'off',
        'implicit-arrow-linebreak': 'off',
        'max-params': 2,
        'no-alert': 'error',
        'no-console': 'error',
        'no-eval': 'error',
        'no-tabs': ['error', { allowIndentationTabs: true }],
        'no-unused-vars': 'off', // prefer @typescript-eslint version
        'prefer-const': 'error',
        'prefer-destructuring': ['error', { array: false, object: true }],

        // jest rules either not in recommended rule set or overridden
        'jest/expect-expect': ['error', { assertFunctionNames: ['expect*'] }],
        'jest/no-mocks-import': 'off',

        // react rules either not in recommended rule set or overridden
        'react/destructuring-assignment': ['error', 'always'],
        'react/jsx-boolean-value': 'error',
        'react/jsx-no-constructed-context-values': ['error'],
        'react/jsx-no-useless-fragment': ['error', { allowExpressions: true }],
        'react/jsx-props-no-spreading': 'off',
        'react/jsx-wrap-multilines': ['error', { declaration: 'ignore' }],
        'react/no-array-index-key': 'off',
        'react/no-danger': ['error'],
        'react/no-unused-prop-types': ['error'],
        'react/prop-types': 'off',
        'react/static-property-placement': ['error', 'static public field'],
      },
    },
  ],

  rules: sharedRules,
};
