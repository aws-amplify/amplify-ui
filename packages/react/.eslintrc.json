{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "extends": ["plugin:react-hooks/recommended"],
  "rules": {
    "react-hooks/rules-of-hooks": "error",
    "react-hooks/exhaustive-deps": "error"
  },
  "parserOptions": {
    "ecmaVersion": "latest",
    "sourceType": "module"
  },
  "ignorePatterns": ["scripts", "dist", "node_modules"]
}
