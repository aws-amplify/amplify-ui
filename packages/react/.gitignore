dist
# ignore generated primitive catalog ESM
src/PrimitiveCatalog.ts 

# ignore jest generated coverage directory
coverage 
