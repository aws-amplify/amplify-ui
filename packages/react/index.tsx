import { I18n } from '@aws-amplify/core';
import { dict } from '../core';

export * from './src';

I18n.putVocabularies(dict);
