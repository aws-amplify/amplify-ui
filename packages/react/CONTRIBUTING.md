# Contributing

## Getting Started

1. Navigate to the _root_ of your local clone of [aws-amplify/amplify-ui](https://github.com/aws-amplify/amplify-ui)
1. Run `yarn setup`
1. Run `yarn react dev`

This will start building `@aws-amplify/ui-react` in watch mode. To test your changes, you can utilize [`examples/next`](../../examples/next) to run examples on `next.js`. Please see examples [README](../../examples/README.md) and e2e [README](../e2e/README.md#contributing) to get started.

## Dependencies

`@aws-amplify/ui-react` depends on [`@aws-amplify/ui`](../ui) for theming, state management, and translation logic. If you're looking for change in these, please refer to `@aws-amplify/ui` [README](../ui/README.md).

## Code Standards

### Imports

Separate all imports into organized, alphabetical blocks of `third_party → internal → local` for easier reading.

```js
import { isEmpty } from 'lodash/isEmpty';
import * as React from 'react';

import { Auth } from 'aws-amplify';

import { Button } from './primitives/Button';
import { THIS_ENUM } from './utils/types';
```

Do **NOT** use implicit paths like below. This can lead to circular dependencies unintentionally which is bad for tree shaking.

```js
import { Flex, Heading } from '../../..';
```

Use explicit import paths instead, the more specific the better.

```js
import { Flex } from '../../../primitives/Flex';
import { Heading } from '../../../primitives/Heading';
```

### Variable Naming

1. Component names should be capitalized.
2. Boolean props should be prefixed with `is` and default to `false`.

### Others

Do **NOT** introduce any [side effects](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free) unless you have to with rational reasons. A "side effect" is defined as code that performs a special behavior when imported, other than exposing one or more exports. An example of this are polyfills, which affect the global scope and usually do not provide an export. For example:

```js
import './polyfill';
```

Any imported file is subject to tree shaking, that means CSS file needs to be added to the side effect list so it will not be unintentionally dropped:

```js
import './style.css';
```

## Testing for Production

After you tested your change, you can run `yarn react build` from monorepo to run build for production.
