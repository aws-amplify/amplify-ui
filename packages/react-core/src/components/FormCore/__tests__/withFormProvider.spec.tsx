import React from 'react';
import { render } from '@testing-library/react';

import { withFormProvider } from '..';

const Form = ({ pretend }: { pretend: string }) => <>{pretend}</>;

describe('withFormProvider', () => {
  it('renders a Provider wrapped Form as expected', () => {
    const WrappedForm = withFormProvider(Form);

    const { container } = render(<WrappedForm pretend="value" />);

    expect(container).toMatchSnapshot();
  });
});
