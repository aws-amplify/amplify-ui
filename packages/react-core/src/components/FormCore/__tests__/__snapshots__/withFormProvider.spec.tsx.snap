// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFormProvider renders a Provider wrapped Form as expected 1`] = `
<div>
  value
</div>
`;
